src/metaopt/algo/space.py

```diff
@@ -0,0 +1,539 @@
+        a variable is `alpha`-likely to be drawn from.
+
+        .. note:: Lower bound is inclusive, upper bound is exclusive.
+
+        """
+        prior_low, prior_high = super(Real, self).interval(alpha)
+        return (max(prior_low, self._low), min(prior_high, self._high))
+
+    def sample(self, n_samples=1, seed=None):
+        """Draw random samples from `prior`.
+
+        .. seealso:: `Dimension.sample`
+
+        """
+        samples = []
+        for _ in range(n_samples):
+            for _ in range(4):
+                sample = super(Real, self).sample(1, seed)
+                if sample[0] not in self:
+                    nice = False
+                    continue
+                nice = True
+                samples.extend(sample)
+                break
+            if not nice:
+                raise ValueError("Improbable bounds: (low={0}, high={1}). "
+                                 "Please make interval larger.".format(self._low, self._high))
+
+        return samples
+
+
+class Integer(Real):
+    """Subclass of `Dimension` for representing integer parameters.
+
+    Attributes
+    ----------
+    name : str
+    type : str
+    prior : `scipy.stats.distributions.rv_generic`
+    shape : tuple
+       See Attributes of `Dimension`.
+    discrete : bool
+       True
+
+    """
+
+    def __init__(self, name, prior, *args, **kwargs):
+        """Search space dimension that can take on integer values.
+
+        Parameters
+        ----------
+        name : str
+        prior : str
+        args : list
+        kwargs : dict
+           See Parameters of `Dimension.__init__`.
+
+        """
+        kwargs['discrete'] = True
+        super(Integer, self).__init__(name, prior, *args, **kwargs)
+
+    def __contains__(self, point):
+        """Check if constraints hold for this `point` of `Dimension`.
+
+        :param x: a parameter corresponding to this `Dimension`.
+        :type x: numeric or array-like
+
+        `Integer` will check whether `point` contains only integers.
+
+        """
+        point_ = numpy.asarray(point)
+        if not numpy.all(numpy.equal(numpy.mod(point_, 1), 0)):
+            return False
+        return super(Integer, self).__contains__(point)
+
+
+class Categorical(Dimension):
+    """Subclass of `Dimension` for representing integer parameters.
+
+    Attributes
+    ----------
+    name : str
+    type : str
+    prior : `scipy.stats.distributions.rv_generic`
+    shape : tuple
+       See Attributes of `Dimension`.
+    discrete : bool
+       True
+    categories : tuple
+       A set of unordered stuff to pick out from, except if enum
+
+    """
+
+    def __init__(self, name, categories, **kwargs):
+        """Search space dimension that can take on categorical values.
+
+        Parameters
+        ----------
+        name : str
+           See Parameters of `Dimension.__init__`.
+        categories : dict or other iterable
```

Comment:
I am not sure because the only condition to check whether a sample has been generated by this dimension type [to be `in self.categories`](https://github.com/tsirif/metaopt/blob/feature/algorithm/src/metaopt/algo/space.py#L380). However I use OrderedDict in tests to keep things deterministic => testable. Orderlessness or indeterministic order should not affect the functionality though in any case.

Reply:
You mean that the categories dictionary is only used with get accesses and never iterated on? I think it'd be safer to assume it could at some point be iterated on, and since users define it in a precise order it could be worth preserving that order.

Reply:
Categories are an unordered set, that's why I thought it would not matter.

Reply:
More or less, I believe users might think the order matter. It might affect the visualisations interface, like keeping the same order.